File: solutions_2453486_1/C++/meisterT42/tttt.cpp
#include <iostream>
#include <vector>

using namespace std;

int main(int argc, char **argv) {
	int tc;
	cin >> tc;
	for (int c = 0; c < tc; c++) {
		cout << "Case #" << (c + 1) << ": ";
		vector<string> board;
		for (int i = 0; i < 4; i++) {
			string s;
			cin >> s;
			board.push_back(s);
		}
		bool completed = true;
		bool Owon = false;
		bool Xwon = false;
		for (int i = 0; i < 4; i++) {
			bool Xposs = true;
			bool Oposs = true;
			for (int j = 0; j < 4; j++) {
				if (board[i][j] == '.') {
					completed = Xposs = Oposs = false;
				} else if (board[i][j] == 'X') {
					Oposs = false;
				} else if (board[i][j] == 'O') {
					Xposs = false;
				}
			}
			if (Xposs) Xwon = true;
			if (Oposs) Owon = true;
			Xposs = true;
			Oposs = true;
			for (int j = 0; j < 4; j++) {
				if (board[j][i] == '.') {
					completed = Xposs = Oposs = false;
				} else if (board[j][i] == 'X') {
					Oposs = false;
				} else if (board[j][i] == 'O') {
					Xposs = false;
				}
			}
			if (Xposs) Xwon = true;
			if (Oposs) Owon = true;
		}
		bool Xposs = true;
		bool Oposs = true;
		for (int j = 0; j < 4; j++) {
			if (board[j][j] == '.') {
				completed = Xposs = Oposs = false;
			} else if (board[j][j] == 'X') {
				Oposs = false;
			} else if (board[j][j] == 'O') {
				Xposs = false;
			}
		}
		if (Xposs) Xwon = true;
		if (Oposs) Owon = true;
		Xposs = true;
		Oposs = true;
		for (int j = 0; j < 4; j++) {
			if (board[3-j][j] == '.') {
				completed = Xposs = Oposs = false;
			} else if (board[3-j][j] == 'X') {
				Oposs = false;
			} else if (board[3-j][j] == 'O') {
				Xposs = false;
			}
		}
		if (Xposs) Xwon = true;
		if (Oposs) Owon = true;
		if (Xwon) cout << "X won" << endl;
		else if (Owon) cout << "O won" << endl;
		else if (!completed) cout << "Game has not completed" << endl;
		else cout << "Draw" << endl;
	}
	return 0;
}

File: solutions_2453486_0/Python/thecook/tttt.py
import sys

#index by row then column
def readboard(f):
    board = []
    for row in range(4):
        line = []
        for c in f.readline():
            if c in ['X', 'O', 'T', '.']:
                line.append(c)
        board.append(line)
    f.readline()
    return board

def isequaltochar(c, other):
    return c == other or 'T' == other
    
def checkforwin(board, c):
    for row in range(4):
        equal = True
        for col in range(4):
            if not isequaltochar(c, board[row][col]):
                equal = False
                break
        if equal:
            return True
    
    for col in range(4):
        equal = True
        for row in range(4):
            if not isequaltochar(c, board[row][col]):
                equal = False
                break
        if equal:
            return True
            
    if isequaltochar(c, board[0][0]) and isequaltochar(c, board[1][1]) and isequaltochar(c, board[2][2]) and isequaltochar(c, board[3][3]):
        return True
        
    if isequaltochar(c, board[0][3]) and isequaltochar(c, board[1][2]) and isequaltochar(c, board[2][1]) and isequaltochar(c, board[3][0]):
        return True
        
    return False
        
    
def checkforemptyspace(board):
    for row in board:
        for c in row:
            if c == '.':
                return True
    return False

if __name__=="__main__":
    filename = sys.argv[1]
    f = open(filename, 'r')
    numtests = int(f.readline())
    for i in range(1, numtests+1):
        board = readboard(f)
        if checkforwin(board, 'X'):
            print "Case #" + str(i) + ": X won"
        elif checkforwin(board, 'O'):
            print "Case #" + str(i) + ": O won"
        elif checkforemptyspace(board):
            print "Case #" + str(i) + ": Game has not completed"
        else:
            print "Case #" + str(i) + ": Draw"